server/bundles/io.cloudbeaver.service.security/src/io/cloudbeaver/service/security/db/H2Migrator.java

```diff
@@ -0,0 +1,219 @@
+import org.jkiss.dbeaver.registry.DataSourceProviderRegistry;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.util.Properties;
+
+public class H2Migrator {
+
+    private static final Log log = Log.getLog(H2Migrator.class);
+
+    private static final String H2_DRIVER_NAME = "h2_embedded";
+    private static final String H2_V2_DRIVER_NAME = "h2_embedded_v2";
+    private static final String BACKUP_V1_FILENAME = "h2db_backup";
+    private static final String EXPORT_FILE_NAME = "H2v1ExportScript";
+    @SuppressWarnings("SqlNoDataSourceInspection")
+    // language=H2
+    private static final String EXPORT_SCRIPT = "SCRIPT TO ? COMPRESSION DEFLATE CIPHER AES PASSWORD ? CHARSET 'UTF-8'";
+    @SuppressWarnings("SqlNoDataSourceInspection")
+    // no 'language=H2' due to an annoying error about unresolvable statement 'FROM_1X'
+    private static final String IMPORT_SCRIPT = "RUNSCRIPT FROM ? COMPRESSION DEFLATE CIPHER AES PASSWORD ? CHARSET 'UTF-8' FROM_1X";
+
+    private final DBRProgressMonitor monitor;
+    private final DataSourceProviderRegistry dataSourceProviderRegistry;
+    private final CBDatabaseConfig databaseConfiguration;
+    private final String dbUrl;
+    private final Properties dbProperties;
+
+    public H2Migrator(
+        @NotNull DBRProgressMonitor monitor,
+        @NotNull DataSourceProviderRegistry dataSourceProviderRegistry,
+        @NotNull CBDatabaseConfig databaseConfiguration,
+        @NotNull String dbUrl,
+        @NotNull Properties dbProperties
+    ) {
+        this.monitor = monitor;
+        this.dataSourceProviderRegistry = dataSourceProviderRegistry;
+        this.databaseConfiguration = databaseConfiguration;
+        this.dbUrl = dbUrl;
+        this.dbProperties = dbProperties;
+    }
+
+    /**
+     * Migrates the H2 database from version 1 to version 2.
+     * <p>
+     * IMPORTANT: if the database is already version 2, but the driver in the configuration is version 1,
+     * this method updates the database configuration instance in memory only.
+     * </p>
+     */
+    public void tryToMigrateDb() {
+        Driver v2Driver = getV2DriverIfMigrationNeeded();
+        if (v2Driver == null) {
+            if (H2_DRIVER_NAME.equals(databaseConfiguration.getDriver())) {
+                updateConfig(); // if it is already migrated H2 v2 we need to update the config
+            }
+            return;
+        }
+
+        var workPaths = new WorkspacePaths(dbUrl);
+        try {
+            migrateDb(workPaths, v2Driver);
+            updateConfig();
+        } catch (Exception e) {
+            rollback(workPaths);
+            log.error("H2 to v2 migration failed and reverted", e);
+        }
+    }
+
+    @Nullable
+    private Driver getV2DriverIfMigrationNeeded() {
```

Comment:
We cannot use the filename to check that the database is v1, because we offer the possibility to customize the filename in the config:
```
server: {
  database: {
    url: "jdbc:h2:${workspace}/.data/cb.h2.dat"
```
At the same time, when trying to connect with an inappropriate driver, we get a well-known error message that only H2 contributors can change.

On the other hand, it may be worth shortening the text with which we determine the connection error:
```
General error: "The write format 1 is smaller than the supported format
```
instead of
```
General error: "The write format 1 is smaller than the supported format 2
```

Reply:
If users customize their database configuration in any maner then we don't need to do any migration, it is on users. Otherwise it becomes too unpredictable and too complicated.

Migration must upgrade only our preconfigured out-of-the-box database.

Reply:
Ok, implemented